Fix #15633: Product creation: EAN13 barcode automatic generation incompatible with EAN13 existing barcode scanning, unusable function.

zuiko opened the pull request:
# Fix  #15633
For the automatic barcode generation function, for products, in the EAN13 mode, the goal is to ensure that :
- Dolibarr recognize its generated EAN13 codes, in product selection, when it has printed them.
- Dolibarr allows, simultaneously, automatic generation of EAN13 codes for some products and manual entry of scanned codes for other products.
- Basically, with this correction, EAN13 codes are always now stored complete with 13 digits in the database (at least for products).

These three points are therefore different from previous versions.

To do this, the syntax of the mask is modified :
- the last character of the mask, in 13th position, after the last } is necessarily * or ?
- `*` or ? will be replaced by the control key via the EAN13 algorithm.
- the rest of the syntax remains unchanged.

The help comments have been modified accordingly in fr-FR and en-US languages (admin.lang files). 
I let the other languages do their own translation. 
The code is only active if the backoffice barcode module choice in order to encode products barcode is EAN13:
`$type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE; //get barcode type configuration for products
		if ($type==2) //2 = EAN13`
I made and tested this correction on V12.0.4.
The wiki will have to be modified because it does not conform to this change.

zuiko committed:
Update mod_barcode_product_standard.php

Fix #15633

zuiko committed:
Update admin.lang

Fix #15633

zuiko committed:
Update admin.lang

stickler-ci committed:
Fixing style errors.

zuiko commented:
I need this fix on V12 currently on production on my side but if the principle of this fix seems OK for everybody I could carry forward this correction on the current development branch. Just tell me...

eldy committed:
Update mod_barcode_product_standard.php

eldy commented on htdocs/core/modules/barcode/mod_barcode_product_standard.php:
```diff
@@ -171,7 +172,22 @@ public function getNextValue($objproduct = null, $type = '')
 		$numFinal = get_next_value($db, $mask, 'product', $field, $where, '', $now);
-
+		//if EAN13 calculate and substitute the last 13th character (* or ?) used in the mask by the EAN13 key
+		$type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE; //get barcode type configuration for products
```
The $type is already defined and passed to the function. It may be different than PRODUIT_DEFAULT_BARCODE_TYPE for example when function is called to getNextValue for something else than an EAN13. Why adding this line ?

zuiko replied:
Hi Eldy,
In fact I also thought that $type was passed to the function but I put a tracer on it and it only returned 0 while I was waiting for EAN information. That's why I decided to test the EAN configuration for the products because that's what automatic generation is used for.

zuiko replied:
Hello,
I am very motivated to make this EAN13 barcode generation function definitively operational, on the one hand because it is indispensable to me and my users, but on the other hand because it's a problem for a significant number of people, some of whom have posted on the forums, particularly in French, for "some time": [example 1](https://www.dolibarr.fr/forum/t/generateur-code-barre/24076/6), [example 2](https://www.dolibarr.fr/forum/t/probleme-de-code-barre/27070), [example 3](https://www.dolibarr.fr/forum/t/generation-automatique-codes-barres-des-produits/20862/3), [example 4](https://www.dolibarr.fr/forum/t/utiliser-code-barre-lors-de-creation-de-proposition-commerciale/32597/3).
Thank you for telling me how I can help further in this direction.

fappels replied:
$type is code field in c_barcode_type (EAN13) coming from commonobject ->fetch_barcode(), not sure why you have null.
Shouldn't be $mask checked for * or ? and not $numFinal?

zuiko replied:
Hi Fappels,
- may be I get NULL because of an upstream bug but I didn't identify it, so to bypass I found an other (but not perfect) solution.
- $mask is fixed and contains (for EAN13) for example 020{000000000}? , $numFinal is the calculated next value from the $mask, following this example 020000000095? , by chance, reading the code and experimenting it, I found that the function get_next_value in the preceding line:
`$numFinal = get_next_value($db, $mask, 'product', $field, $where, '', $now);`
if the last character is only * or ? (fonctionning as jokers), can calculate the next value on the first 12 digits of the mask, leaving the last character (* or ?) unchanged. So you have just to test it and replace it by the EAN13 key (which calculation function barcode_gen_ean_sum already exists on /core/lib/barcode.lib.php.
That's what my code does.

zuiko replied:
@fappels, @eldy 
it seems that public function fetch_barcode() is not called for the barcode automatic calculation even to give the next value on the codebarre module settings, nor to give the next value on the product codebarre field. 
It is only called to show the codebarre type on the product type field.
So it could explain why $type is null/empty on public function getNextValue of mod_barcode_product_standard.php if it is never passed...
For the product I can easily improve this in htdocs/product/card.php
I will propose you soon an improved code, replacing
`$tmpcode = $modBarCodeProduct->getNextValue($object, $type);`
by
`$tmpcode = $modBarCodeProduct->getNextValue($object, $fk_barcode_type);`
and testing it consequently.

eldy replied:
> Hi Eldy,
> In fact I also thought that $type was passed to the function but I put a tracer on it and it only returned null while I was waiting for EAN information (it's true setting a product barcode and showing the next generated barcode in the module config). That's why I decided to test the EAN configuration for the products because that's what automatic generation is used for.

So if it is null, there is a problem with caller. You can add a workaround to fix this but you must make
$type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE; 
only if it is not already defined.

zuiko committed:
Update mod_barcode_product_standard.php

get barcode type configuration for products only if $type not set

zuiko committed:
Update card.php

$type replaced by $fk_barcode_type in
$tmpcode = $modBarCodeProduct->getNextValue($object, $type);
in order to set it

zuiko commented:
@fappels, @eldy
$fk_barcode_type instead of empty $type has been passed to getNextValue in order to be used in the function everywhere in card.php (products) for barcode purpose.
I couldn't find a function to translate and test textually EAN13 instead of the numerical value 2 in the function getNextValue.
Is EAN13 = 2 a perennial match? 
This is a quite important question for the functioning of my fix over the releases.

It remains a conflict with current dev on htdocs/product/card.php, I can't resolve it alone.

fappels commented:
Did not try, but think if call $object->fetch_barcode() before the 3 $modBarCodeProduct->getNextValue's in product card and put $object->barcode_type_code as second getNextvalue parameter you can check on 'EAN13' iso a database rowid.

zuiko commented:
@fappels 
It could be a good idea, but I fear it does not cover when there is no barcode type and you can see it's replaced by the configuration's one with this instruction :
`$fk_barcode_type = $conf->global->PRODUIT_DEFAULT_BARCODE_TYPE;`
in htdocs/product/card.php and also in my proposal  correction in getNextValue
I think that the perfect solution could be to find a way to translate the numerical barcode type into the litteral matching directly in getNextValue in order to test it.

PS: since the argument $object is passed to the function getNextValue we might think we could use your suggestion directly in my code in getNextvalue function but there will be always the case where $type will be null where you need to replace it by $conf->global->PRODUIT_DEFAULT_BARCODE_TYPE...

fappels commented:
I see in fetch_barcode() this:
$idtype = $this->barcode_type;
if (empty($idtype) && $idtype != '0')	// If type of barcode no set, we try to guess. If set to '0' it means we forced to have type remain not defined
{
	if ($this->element == 'product')      $idtype = $conf->global->PRODUIT_DEFAULT_BARCODE_TYPE;
        elseif ($this->element == 'societe') $idtype = $conf->global->GENBARCODE_BARCODETYPE_THIRDPARTY;
	else dol_syslog('Call fetch_barcode with barcode_type not defined and cant be guessed', LOG_WARNING);
}
So think 'no barcode type' is covered

zuiko commented:
I can confirm that:
```
               $objproduct->fetch_barcode();
     	       $typelitteral=$objproduct->barcode_type_code;
		 echo ('typelitteral = '.$typelitteral);
```
in getNextValue print "EAN13" if $objproduct is passed to the function.
So, if you think it's always the case and if now $type is always different of null after my fix, I could test EAN13 in litteral without replacing it by $conf->global->PRODUIT_DEFAULT_BARCODE_TYPE in this function.

Do you confirm ?

fappels commented:
Yes you should be able to test in literal, but best test by edit and create new product. There are 3 $modBarCodeProduct->getNextValue in product card, so best test three situation.

zuiko commented:
In fact, testing, I succeded only to pass in 
`$tmpcode = $modBarCodeProduct->getNextValue($object, $fk_barcode_type);`
only in the line around 1833 in card.php.
I don't know (and not search) exactly how to pass thru the 2 other preceding cases.

But if I put:  
```
$objproduct->fetch_barcode();
$typelitteral=$objproduct->barcode_type_code;
```
directly in getNextValue I thought I could cover the 3 cases in one time, isn't it ?
The last question is: will it cover the next barcode value showed in the barcode module settings itself?...

zuiko commented:
_**"The last question is: will it cover the next barcode value showed in the barcode module settings itself?..."**_
I confirm that $objproduct is not passed to getNextValue in that case, so, we have to find a solution for this specific case, to translate $type=$conf->global->PRODUIT_DEFAULT_BARCODE_TYPE to literal value...

/hdtdocs/admin/barcode.php 
`print '<td class="nowrap">'.$modBarCode->getExample($langs)."</td>\n";`

/htdocs/core/modules/barcode/mod_barcode_product_standard.php
```

	public function getExample($langs, $objproduct = 0)
	{
		$examplebarcode = $this->getNextValue($objproduct, '');
```

eldy committed:
Update mod_barcode_product_standard.php

stickler-ci committed:
Fixing style errors.

zuiko commented:
OK @eldy 
your update is cleaner than mine.
And, of course, it works, I tested it.

eldy committed:
Merge branch 'develop' into patch-1

stickler-ci committed:
Fixing style errors.

zuiko committed:
Update mod_barcode_product_standard.php

More evolving coding by testing the type of barcode in literal form.
Does not change deeply the #15918 correction principle in order to fix bug #15633.

stickler-ci committed:
Fixing style errors.

zuiko commented:
Hi @eldy and @fappels 
I just updated mod_barcode_product_standard.php
in order to propose a more evolving coding by testing the type of barcode in literal form.
For the literal form, in literalBarcodeType, at first I took the barcode code by this instruction:
`$out .= $obj->code;`
but I hesitated and I think it's better to have taken the libelle by this replacing instruction:
`$out .= $obj->libelle;`
For EAN13 it's the same result but it's not the case for all barcodes:
rowid	code	libelle
	1 	EAN8 	EAN8
	2 	EAN13 	EAN13
	3 	UPC 	UPC
	4 	ISBN 	ISBN
	5 	C39 	Code 39
	6 	C128 	Code 128
	7 	DATAMATRIX 	Datamatrix
	8 	QRCODE 	Qr Code
and I think, it's less ambiguous for everybody to test what is written directly on screen, so I will change soon.
All of this does not change deeply the #15918 correction principle in order to fix bug #15633.

zuiko committed:
Update mod_barcode_product_standard.php

Blank line after comment removed...

zuiko committed:
Update mod_barcode_product_standard.php

$out .= $obj->libelle; //take the libelle corresponding to the type rowid in the database
Take the libelle instead the code for the litteral value of the barcode type rowid.

eldy committed:
Debug

stickler-ci committed:
Fixing style errors.

zuiko commented:
Thanks for the debug @eldy,
in fact I built the public function literalBarcodeType code the nearer possible than public function selectBarcodeType I found in /htdocs/core/class/html.formbarcode.class.php

_(And, of course, your code works, I tested it.)_

zuiko commented:
Just a beginner question about Dolibarr project :
- who is in charge to transfer this code on the last development branch?

eldy commented:
Merge into develop is done by myself.

eldy commented on htdocs/core/modules/barcode/mod_barcode_product_standard.php:
```diff
@@ -171,7 +206,23 @@ public function getNextValue($objproduct = null, $type = '')
+		//if EAN13 calculate and substitute the last 13th character (* or ?) used in the mask by the EAN13 key
+		$literaltype = '';
+		$literaltype = $this->literalBarcodeType($db, $type);//get literal_Barcode_Type
+		if ($literaltype=='EAN13') //EAN13 rowid = 2
```
The call $this->literalBarcodeType and test on == 'EAN13) should be done after the 
if ((substr($numFinal, -1)=='*') or (substr($numFinal, -1)=='?')
So we make a sql request only if there is the ? at end of string. This will save a sql when type is not ean13

zuiko replied:
ok with @eldy
I sent an other code proposal:
Avoiding  sql call if no joker at the end of mask
Treating EAN13 in a more evolutive code form to prepare needs for other barcodes with key (if any).

zuiko committed:
Update mod_barcode_product_standard.php

Fix #15633: Product creation: EAN13 barcode automatic generation incompatible with EAN13 existing barcode scanning, unusable function.
#15918 
ok with @eldy
Avoid  sql call if no joker at the end of mask
Treat EAN13 in a more evolutive code form to prepare needs for other barcodes with key

stickler-ci committed:
Fixing style errors.
